impl.go

```diff
@@ -111,10 +166,197 @@ func findInterface(iface string, srcDir string) (path string, id string, err err
 	}
 	decl := f.Decls[1].(*ast.GenDecl)      // var i io.Reader
 	spec := decl.Specs[0].(*ast.ValueSpec) // i io.Reader
-	sel := spec.Type.(*ast.SelectorExpr)   // io.Reader
-	id = sel.Sel.Name                      // Reader
+	iface, err = typeFromAST(spec.Type)
+	if err != nil {
+		return path, iface, fmt.Errorf("error parsing type from AST: %w", err)
+	}
+	// trim off the package which got smooshed on when resolving the type
+	_, iface.ID, _ = strings.Cut(iface.ID, ".")
+	return path, iface, err
+}
+
+func typeFromAST(in ast.Expr) (Type, error) {
+	switch specType := in.(type) {
+	case *ast.Ident:
+		// a standalone identifier (Reader) shows up as an Ident
+		return Type{ID: specType.Name}, nil
+	case *ast.SelectorExpr:
+		// an identifier in a different package (io.Reader) shows up as a SelectorExpr
+		// we need to pull the name out
+		return Type{ID: specType.X.(*ast.Ident).Name + "." + specType.Sel.Name}, nil
+	case *ast.StarExpr:
+		// pointer identifiers (*Reader) show up as a StarExpr
+		// we need to pull the name out and prefix it with a *
+		typ, err := typeFromAST(specType.X)
+		if err != nil {
+			return Type{}, err
+		}
+		typ.ID = "*" + typ.ID
+		return typ, nil
+	case *ast.ArrayType:
+		// slices and arrays ([]Reader) show up as an ArrayType
+		typ, err := typeFromAST(specType.Elt)
+		if err != nil {
+			return Type{}, err
+		}
+		prefix := "["
+		if specType.Len != nil {
+			prefix += specType.Len.(*ast.BasicLit).Value
+		}
+		prefix += "]"
+		typ.ID = prefix + typ.ID
+		return typ, nil
+	case *ast.MapType:
+		// maps (map[string]Reader) show up as a MapType
+		key, err := typeFromAST(specType.Key)
+		if err != nil {
+			return Type{}, err
+		}
+		value, err := typeFromAST(specType.Value)
+		if err != nil {
+			return Type{}, err
+		}
+		return Type{
+			ID: "map[" + key.String() + "]" + value.String(),
```

Comment:
in other places you're careful to only touch the ID, not the params, of subtypes but here, the params get clobbered. makes me feel unsure i understand what this code is doing.